zip_0243.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+from pyblake2 import blake2b
+import struct
+
+from transaction import (
+    MAX_MONEY,
+    SAPLING_TX_VERSION,
+    Script,
+    Transaction,
+)
+from tv_output import render_args, render_tv, Some
+from tv_rand import Rand
+
+from zip_0143 import (
+    getHashJoinSplits,
+    getHashOutputs,
+    getHashPrevouts,
+    getHashSequence,
+    NOT_AN_INPUT,
+    SIGHASH_ALL,
+    SIGHASH_ANYONECANPAY,
+    SIGHASH_NONE,
+    SIGHASH_SINGLE,
+)
+
+
+def getHashShieldedSpends(tx):
+    digest = blake2b(digest_size=32, person=b'ZcashSSpendsHash')
+    for desc in tx.vShieldedSpends:
+        # We don't pass in serialized form of desc as spendAuthSig is not part of the hash
+        digest.update(bytes(desc.cv))
+        digest.update(bytes(desc.anchor))
+        digest.update(desc.nullifier)
+        digest.update(desc.rk)
+        digest.update(bytes(desc.proof))
+    return digest.digest()
+
+def getHashShieldedOutputs(tx):
+    digest = blake2b(digest_size=32, person=b'ZcashSOutputHash')
+    for desc in tx.vShieldedOutputs:
+        digest.update(bytes(desc))
+    return digest.digest()
+
+def signature_hash(scriptCode, tx, nIn, nHashType, amount, consensusBranchId):
```

Comment:
Wonder if there's a way to do this pr with less code duplication - only the two lines about hashes of spends and outputs are new.

Reply:
A good idea to refactor this and reduce code duplication.  Something for post-Sapling activation.